src/Moq/Extensions.cs

```diff
@@ -369,6 +369,94 @@ private static MethodInfo GetInvokeMethodFromUntypedDelegateCallback(Delegate ca
 			}
 		}
 
+		/// <summary>
+		///   Visits all constituent parts of <paramref name="type"/>, replacing all type matchers
+		///   that match the type argument at the corresponding position in <paramref name="other"/>.
+		/// </summary>
+		/// <param name="type">The type to be matched. May be, or contain, type matchers.</param>
+		/// <param name="other">The type argument to match against <paramref name="type"/>.</param>
+		public static Type SubstituteTypeMatchers(this Type type, Type other)
+		{
+			// If a type matcher `T` successfully matches its corresponding type `O` from `other`, `T` in `type`
+			// gets replaced by `O`. If all type matchers successfully matched, and they have been replaced by
+			// their arguments in `other`, callers can then perform a final `IsAssignableFrom` check to match
+			// everything else (fixed types). Being able to defer to `IsAssignableFrom` saves us from having to
+			// re-implement all of its type equivalence rules (polymorphism, co-/contravariance, etc.).
+			//
+			// We still need to do some checks ourselves, however: In order to traverse both `type` and `other`
+			// in lockstep, we need to ensure that they have the same basic structure.
+
+			if (type.IsTypeMatcher(out var typeMatcherType))
+			{
+				var typeMatcher = (ITypeMatcher)Activator.CreateInstance(typeMatcherType);
+
+				if (typeMatcher.Matches(other))
+				{
+					return other;
+				}
+			}
+			else if (type.HasElementType && other.HasElementType)
+			{
+				var te = type.GetElementType();
+				var oe = other.GetElementType();
+
+				if (type.IsArray && other.IsArray)
+				{
+					var tr = type.GetArrayRank();
+					var or = other.GetArrayRank();
+
+					if (tr == or)
+					{
+						var se = te.SubstituteTypeMatchers(oe);
+						if (se == te)
+						{
+							return type;
+						}
+						else
+						{
+							return tr == 1 ? se.MakeArrayType() : se.MakeArrayType(tr);
+						}
+					}
+				}
+				else if (type.IsByRef && other.IsByRef)
+				{
+					var se = te.SubstituteTypeMatchers(oe);
+					return se == te ? type : se.MakeByRefType();
+				}
+				else if (type.IsPointer && other.IsPointer)
+				{
+					var se = te.SubstituteTypeMatchers(oe);
+					return se == te ? type : se.MakePointerType();
+				}
+			}
+			else if (type.IsGenericType && other.IsGenericType)
+			{
+				var td = type.GetGenericTypeDefinition();
+				var od = other.GetGenericTypeDefinition();
+
+				if (td == od)
```

Comment:
Never quite sure about the runtime's guarantees re: reference equality of `Type` instances; so `.Equals()` might be safer than `==`.